File: ex/day1.ex
defmodule String.Extra do
  def substrings(str, size) do
    0..(String.length(str) - 1)
    |> Enum.map(&String.slice(str, &1..(&1 + size)))
  end
end

defmodule Day1 do
  def filter_to_integers(line) do
    line
    |> String.graphemes()
    |> Enum.filter(fn char ->
      case Integer.parse(char) do
        {_, ""} -> true
        _ -> false
      end
    end)
  end

  def concatenate_first_and_last(list) do
    List.first(list) <> List.last(list)
  end

  def solve_part1(input) do
    input
    |> String.split("\n")
    |> Enum.map(&Day1.filter_to_integers/1)
    |> Enum.map(fn integers ->
      integers
      |> Day1.concatenate_first_and_last()
      |> String.to_integer()
    end)
    |> Enum.sum()
  end

  def parse_text_to_number(s) do
    text_to_numbers = %{
      "one" => "1",
      "two" => "2",
      "three" => "3",
      "four" => "4",
      "five" => "5",
      "six" => "6",
      "seven" => "7",
      "eight" => "8",
      "nine" => "9"
    }

    case Integer.parse(String.first(s)) do
      {_, ""} ->
        String.first(s)

      :error ->
        text_to_numbers
        |> Enum.find(
          {"", ""},
          fn {text, _} -> String.starts_with?(s, text) end
        )
        |> elem(1)
    end
  end

  def solve_part2(input) do
    input
    |> String.split("\n")
    |> Enum.map(fn integers ->
      integers
      |> String.Extra.substrings(5)
      |> Enum.map(&Day1.parse_text_to_number/1)
    end)
    |> Enum.map(fn integers ->
      integers
      |> Enum.filter(fn x -> x !== "" end)
      |> Day1.concatenate_first_and_last()
      |> String.to_integer()
    end)
    |> Enum.sum()
  end
end

File: rs/src/bin/day1.rs
use clap::Parser;
use regex::Regex;

#[derive(Parser)]
struct Cli {
    path: std::path::PathBuf,
}

fn find_digits(line: &str) -> Vec<i32> {
    let re = Regex::new(r"\d").unwrap();

    let digits: Vec<_> = re.find_iter(line)
        .map(|d| d.as_str().parse::<i32>().unwrap())
        .collect();

    return digits;
}

fn combine_first_and_last(digits: Vec<i32>) -> i32 {
    return 10 * digits.first().unwrap() + digits.last().unwrap();
}

fn solve_part1(content: String) -> i32 {
    return content
        .lines()
        .map(|line| {
        let digits = find_digits(line);
        let combined = combine_first_and_last(digits);
        return combined;
    }).sum()
}

fn solve_part2(content: String) -> i32 {
    return content
        .lines()
        .map(|line| {
        let updated_line = line
            .replace("one", "one1one")
            .replace("two", "two2two")
            .replace("three", "three3three")
            .replace("four", "four4four")
            .replace("five", "five5five")
            .replace("six", "six6six")
            .replace("seven", "seven7seven")
            .replace("eight", "eight8eight")
            .replace("nine", "nine9nine");
        let digits = find_digits(&updated_line);
        let combined = combine_first_and_last(digits);
        return combined;
    }).sum()
}

fn main() {
    let args = Cli::parse();
    println!("input: {:?}", args.path);

    let content = std::fs::read_to_string(&args.path).expect("could not read file");

    println!("part1: {}", solve_part1(content.clone()));
    println!("part2: {}", solve_part2(content.clone()));
}
